src/code/z_fireobj.c

```diff
@@ -1,27 +1,244 @@
 #include "global.h"
+#include "overlays/actors/ovl_En_Arrow/z_en_arrow.h"
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_fireobj/FireObj_InitWithParams.s")
+ColliderCylinderInit fireObjCollisionInit = {
+    {
+        COLTYPE_NONE,
+        AT_NONE,
+        AC_ON | AC_TYPE_PLAYER,
+        OC1_NONE,
+        OC2_NONE,
+        COLSHAPE_CYLINDER,
+    },
+    {
+        ELEMTYPE_UNK4,
+        { 0x00000000, 0x00, 0x00 },
+        { 0x00000820, 0x00, 0x00 },
+        TOUCH_NONE | TOUCH_SFX_NORMAL,
+        BUMP_ON,
+        OCELEM_NONE,
+    },
+    { 18, 67, 0, { 0, 0, 0 } },
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_fireobj/FireObj_SetState.s")
+FireObjLightParams sFireObjLightParams[2] = {
+    { 0xFF, 0xFF, 0x00, 0xFF, 0x80, 0xFF, 0x00, 0x00 },
+    { 0x00, 0xC8, 0x80, 0x80, 0x00, 0x7F, 0x7F, 0x00 },
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_fireobj/FireObj_SetPosition.s")
+void FireObj_InitLight(GlobalContext* globalCtx, FireObjLight* light, u8* param_3, Vec3f* pos);
+void FireObj_UpdateLight(GlobalContext* globalCtx, FireObjLight* light, FireObj* fire);
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_fireobj/FireObj_StepSize.s")
+void FireObj_InitWithParams(GlobalContext* globalCtx, FireObj* fire, FireObjInitParams* init) {
+    fire->size = init->size;
+    fire->sizeInv = 1.0f / init->size;
+    fire->dynamicSizeStep = init->dynamicSizeStep;
+    fire->state = init->state;
+    fire->sizeGrowsCos2 = init->sizeGrowsCos2;
+    fire->unk27 = init->unkA;
+    fire->flags = init->flags;
+    fire->xScale = 0.0f;
+    fire->yScale = 0.0f;
+    fire->dynamicSize = 0.0f;
+    fire->unk26 = Rand_ZeroOne() * 20.0f;
+    fire->ignitionDelay = -1;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_fireobj/FireObj_UpdateStateTransitions.s")
+void FireObj_SetState(FireObj* fire, f32 dynamicSizeStep, u8 newState) {
+    fire->state = newState & 0xFF;
+    if (fire->state == 3) {
```

Comment:
States should have an enum. For naming the values it seems like there may be enough in this file to surmise what the different states do, but if not even an enum like `FIREOBJ_STATE_0` etc. is better than nothing.